File: apps/android/feature/settings/src/main/kotlin/app/storyarc/feature/settings/WhatsNewSheet.kt
package app.storyarc.feature.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.ListItemDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.SheetValue
import androidx.compose.material3.rememberBottomSheetState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcSpace

/**
 * What changed in the version just installed, shown once and then dismissed.
 *
 * **A `ModalBottomSheet`, not a full-screen dialog.** Material reserves those for a
 * multi-step task holding unsaved state, at compact widths *only*, and says to use a dialog
 * or a side sheet on a tablet instead — StoryArc runs on tablets, and this is neither
 * multi-step nor unsaved. The modal sheet's own guidance describes this content almost word
 * for word: use one *"when items require longer descriptions and icons"*. `design.md` settles
 * it, and iOS's large sheet is deliberately a different shape; [ADR-0001] is what makes that
 * allowed rather than an inconsistency.
 *
 * Capped by Material's own maximum height — it never covers the status bar — and draggable,
 * which is the component's own behaviour and is left alone. Nothing waits on it: the action
 * dismisses it, so does a drag down, so does the scrim, and so does back.
 *
 * **`skipPartiallyExpanded`, and a screenshot is why.** The first version left the partial
 * detent in place, as `design.md`'s "capped and expandable" reads at first glance. A modal
 * sheet lays its content out at the full height and is *translated* down to the partial
 * offset, so the pinned `Continue` sat below the visible edge: the emulator capture showed the
 * fourth row running off the bottom of the screen with no action anywhere. That is the one
 * thing `settings-and-about` says outright — "the dismissing action stays reachable without
 * scrolling past the content" — and no unit test can see it, because the height that hid the
 * button belongs to the dialog rather than to the content. The sheet opens expanded now, which
 * is also the shape iOS's single `.large` detent has.
 */
@OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@Composable
fun WhatsNewSheet(release: WhatsNewRelease, onDismiss: () -> Unit) {
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        // `rememberBottomSheetState` with two values rather than the deprecated
        // `rememberModalBottomSheetState(skipPartiallyExpanded = true)`, which 1.5.0's own
        // KDoc names as the replacement. Hidden and Expanded and nothing between them.
        sheetState = rememberBottomSheetState(
            SheetValue.Hidden,
            setOf(SheetValue.Hidden, SheetValue.Expanded),
        ),
    ) {
        WhatsNewContent(release = release, onDismiss = onDismiss)
    }
}

/**
 * The sheet's contents, which is where every claim about it actually lives.
 *
 * Its own composable because `ModalBottomSheet` is a dialog window and a unit test cannot
 * compose one — `WhatsNewLayoutTest` composes this, in a box the size of the smallest window
 * this app supports, which is the same bound the sheet imposes.
 *
 * **The rows scroll and the action does not.** `settings-and-about` at the largest text size:
 * "the screen scrolls if it must, and the dismissing action stays reachable without scrolling
 * past the content". `weight(1f, fill = false)` is what holds it: the scrolling column takes
 * the height it needs up to whatever is left after the button, and no more.
 */
@Composable
internal fun WhatsNewContent(
    release: WhatsNewRelease,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val palette = LocalStoryArcPalette.current

    Column(modifier = modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .weight(1f, fill = false)
                .verticalScroll(rememberScrollState()),
        ) {
            Column(
                modifier = Modifier.padding(
                    start = StoryArcSpace.gutter,
                    end = StoryArcSpace.gutter,
                    bottom = StoryArcSpace.md,
                ),
                verticalArrangement = Arrangement.spacedBy(StoryArcSpace.xs),
            ) {
                Text(
                    text = stringResource(R.string.whats_new_title),
                    style = MaterialTheme.typography.headlineMedium,
                    color = palette.textPrimary,
                )
                Text(
                    text = stringResource(R.string.whats_new_version, release.version),
                    style = MaterialTheme.typography.bodyMedium,
                    color = palette.textSecondary,
                )
            }
            WhatsNewNotes(release.notes)
        }

        Button(
            onClick = onDismiss,
            modifier = Modifier
                .fillMaxWidth()
                .padding(StoryArcSpace.gutter),
        ) {
            Text(stringResource(R.string.whats_new_continue))
        }
    }
}

/**
 * The same rows, every release of them, reached from About.
 *
 * `settings-and-about`: what changed "is reachable from the About screen, along with the
 * entries for earlier versions", **and** "reaching it that way does not change what the app
 * considers seen". The second half is why this takes a list of releases and nothing else:
 * there is no store here to write to.
 *
 * No scroll of its own, and no `Continue`. `SettingsScreen` already scrolls every group, and
 * a second vertical scroll inside the first is measured with an infinite height — which threw
 * and took the whole app down every time anyone opened About, which is why `AboutGroup` says
 * the same thing at the same length.
 */
@Composable
internal fun WhatsNewHistory(
    releases: List<WhatsNewRelease>,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val palette = LocalStoryArcPalette.current

    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(StoryArcSpace.md),
    ) {
        OutlinedButton(onClick = onBack) { Text(stringResource(R.string.settings_back)) }
        Text(
            text = stringResource(R.string.whats_new_title),
            style = MaterialTheme.typography.titleMedium,
            color = palette.textPrimary,
        )
        for (release in releases) {
            Text(
                text = stringResource(R.string.whats_new_version, release.version),
                style = MaterialTheme.typography.labelLarge,
                color = palette.textTertiary,
                modifier = Modifier.padding(top = StoryArcSpace.sm),
            )
            WhatsNewNotes(release.notes)
        }
    }
}

/** One release's rows, so the sheet and the About screen cannot draw them differently. */
@Composable
private fun WhatsNewNotes(notes: List<WhatsNewNote>) {
    val palette = LocalStoryArcPalette.current

    for (note in notes) {
        ListItem(
            supportingContent = {
                Text(
                    text = stringResource(note.body),
                    style = MaterialTheme.typography.bodyMedium,
                    color = palette.textSecondary,
                )
            },
            leadingContent = {
                // **A fixed column, in dp, which is the point of it.** Material asks for 200%
                // text support and, in the same breath, says not to resize a component that
                // contains no text — and an icon contains none. Left to scale, it takes a
                // third of a 320 dp line at 200% and the sentence beside it is measured into
                // what is left. `WhatsNewLayoutTest` is what holds this, by measuring where
                // the sentence starts at both text sizes.
                //
                // A `Box` around the `Icon` rather than a size on the `Icon` itself: the
                // alignment is what keeps a short icon centred in the column beside a
                // sentence that has grown to four lines.
                Box(
                    modifier = Modifier.size(ICON_COLUMN),
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        imageVector = note.icon,
                        // Decorative. The heading beside it says the same thing in words, and
                        // a screen reader announcing both would say everything twice.
                        contentDescription = null,
                        tint = palette.accent,
                        modifier = Modifier.size(ICON),
                    )
                }
            },
            // Transparent, because the sheet's own container is the surface here and a
            // second one behind every row draws a card the design does not have.
            colors = ListItemDefaults.colors(containerColor = Color.Transparent),
        ) {
            Text(
                text = stringResource(note.title),
                style = MaterialTheme.typography.titleSmall,
                color = palette.textPrimary,
            )
        }
    }
}

/** The width the icons sit in, whatever the reader's text size. */
private val ICON_COLUMN = 40.dp

/** The icon inside it. */
private val ICON = 28.dp

File: apps/ios/Packages/StoryArcKit/Sources/SettingsFeature/WhatsNewSheet.swift
public import SwiftUI

internal import DesignSystem

/// What changed in the version just installed, shown once and then dismissed.
///
/// **Drawn the way Apple draws it**, which is what the owner asked for: a large heading, a
/// short column of rows each carrying an icon, a heading and one sentence, and a single
/// action pinned at the foot. Four rows rather than a changelog — a reader who opens a
/// reading app is there to read, and `settings-and-about` says this "SHALL never let that
/// get in the way".
///
/// **The action is `Continue`, not `Done`, and it is the only one.** The requirement asks
/// that "one action dismisses it", and a sheet with a Continue *and* a Close is two ways to
/// do the same thing. Swiping it down is the platform's own second way and is not ours to
/// remove; it costs nothing here because the version was recorded before this was drawn —
/// see ``WhatsNew/onLaunch(store:)``.
///
/// **The rows scroll and the action does not.** `settings-and-about` at the largest
/// accessibility text size: "the screen scrolls if it must, and the dismissing action stays
/// reachable without scrolling past the content". A `safeAreaInset` is what holds that: the
/// button is outside the scroll view and the scroll view is inset by its height, so nothing
/// is drawn under it either.
public struct WhatsNewSheet: View {
    @Environment(\.theme) private var theme

    private let release: WhatsNewRelease
    private let onDismiss: () -> Void

    public init(release: WhatsNewRelease, onDismiss: @escaping () -> Void) {
        self.release = release
        self.onDismiss = onDismiss
    }

    public var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: StoryArcSpace.xl) {
                VStack(alignment: .leading, spacing: StoryArcSpace.xs) {
                    Text("whatsnew.title", bundle: .module)
                        .textRole(.display)
                        .foregroundStyle(theme.palette.textPrimary)
                    Text("whatsnew.version \(release.version)", bundle: .module)
                        .textRole(.subheadline)
                        .foregroundStyle(theme.palette.textSecondary)
                }
                WhatsNewNotes(notes: release.notes)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, StoryArcSpace.gutter)
            .padding(.top, StoryArcSpace.xxl)
            .padding(.bottom, StoryArcSpace.lg)
        }
        .background(theme.palette.surfaceCanvas)
        .safeAreaInset(edge: .bottom) {
            Button(action: onDismiss) {
                Text("whatsnew.continue", bundle: .module)
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)
            .padding(.horizontal, StoryArcSpace.gutter)
            .padding(.vertical, StoryArcSpace.lg)
            .background(.bar)
        }
    }
}

/// The same rows, every release of them, reached from About.
///
/// `settings-and-about`: what changed "is reachable from the About screen, along with the
/// entries for earlier versions", **and** "reaching it that way does not change what the app
/// considers seen". The second half is why this view takes a list of releases and nothing
/// else: there is no store here to write to. ``WhatsNewWiringTests`` is the tripwire that
/// keeps one from arriving.
///
/// One screen rather than a list of versions leading to a screen each. Two of these releases
/// will fit on a phone for years, and a reader who came here to re-read one sentence should
/// not have to guess which version it was in.
public struct WhatsNewHistory: View {
    @Environment(\.theme) private var theme

    private let releases: [WhatsNewRelease]

    public init(releases: [WhatsNewRelease] = WhatsNew.releases) {
        self.releases = releases
    }

    public var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: StoryArcSpace.xxl) {
                ForEach(releases) { release in
                    VStack(alignment: .leading, spacing: StoryArcSpace.lg) {
                        Text("whatsnew.version \(release.version)", bundle: .module)
                            .textRole(.title3)
                            .foregroundStyle(theme.palette.textPrimary)
                        WhatsNewNotes(notes: release.notes)
                    }
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(StoryArcSpace.gutter)
        }
        .background(theme.palette.surfaceCanvas)
        .navigationTitle(Text("whatsnew.about", bundle: .module))
    }
}

/// One release's rows, so the sheet and the About screen cannot draw them differently.
private struct WhatsNewNotes: View {
    @Environment(\.theme) private var theme

    let notes: [WhatsNewNote]

    var body: some View {
        VStack(alignment: .leading, spacing: StoryArcSpace.xl) {
            ForEach(notes) { note in
                HStack(alignment: .top, spacing: StoryArcSpace.lg) {
                    // **A fixed column, which is the point of it.** The symbol carries no
                    // text, so it has nothing to say at a larger text size — and at the
                    // largest one a scaling icon takes a third of the line and squeezes the
                    // sentence beside it into a column of syllables. Android's row holds the
                    // same width in dp for the same reason, which Material states outright:
                    // do not resize a component that contains no text.
                    Image(systemName: note.symbolName)
                        .font(.system(size: iconPointSize))
                        .foregroundStyle(theme.palette.accent)
                        .frame(width: iconColumnWidth, alignment: .center)
                        .accessibilityHidden(true)
                    VStack(alignment: .leading, spacing: StoryArcSpace.hair) {
                        Text(note.title, bundle: .module)
                            .textRole(.headline)
                            .foregroundStyle(theme.palette.textPrimary)
                        Text(note.body, bundle: .module)
                            .textRole(.subheadline)
                            .foregroundStyle(theme.palette.textSecondary)
                    }
                    .fixedSize(horizontal: false, vertical: true)
                }
            }
        }
    }

    /// Wide enough for the widest symbol drawn at ``iconPointSize``, and no wider.
    private var iconColumnWidth: CGFloat { 34 }
    private var iconPointSize: CGFloat { 26 }
}
